provider/azurerm: Add resource azurerm_sql_elasticpool

dominik-lekse opened the pull request:
This pull request adds the resource `azurerm_sql_elasticpool` to create and manage [Azure SQL Elastic Pools](https://docs.microsoft.com/en-us/azure/sql-database/sql-database-elastic-pool).

Notes:

- In contrast to the existing resources to manage Azure SQL related functionality, `azurerm_sql_elasticpool` uses the `azure-sdk-for-go`. For this purpose, I have added it to the vendor dependencies.

Related issues and pull requests:

- This pull request will allow https://github.com/hashicorp/terraform/pull/12678 to be completed

Acceptance tests:

```
TF_ACC=1 go test ./builtin/providers/azurerm -v -run=TestAccAzureRMSqlElasticPool_ -timeout 120m
=== RUN   TestAccAzureRMSqlElasticPool_importBasic
--- PASS: TestAccAzureRMSqlElasticPool_importBasic (146.96s)
=== RUN   TestAccAzureRMSqlElasticPool_basic
--- PASS: TestAccAzureRMSqlElasticPool_basic (145.20s)
=== RUN   TestAccAzureRMSqlElasticPool_resizeDtu
--- PASS: TestAccAzureRMSqlElasticPool_resizeDtu (187.22s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/azurerm	479.408s
```

Tasks:

- [x] Resource `azurerm_sql_elasticpool`
- [x] Acceptance test `TestAccAzureRMSqlElasticPool_basic` (verifies creating a SQL Elasticpool)
- [x] Acceptance test `TestAccAzureRMSqlElasticPool_resizeDtu` (verifies resizing the DTU limit of a SQL Elasticpool)
- [x] Import acceptance test `TestAccAzureRMSqlElasticPool_importBasic` (verifies a SQL Elasticpool can be imported)
- [x] Documentation
- [ ] Review

tombuildsstuff commented on builtin/providers/azurerm/config.go:
```diff
@@ -99,6 +100,11 @@ type ArmClient struct {
 
 	keyVaultClient keyvault.VaultsClient
+
+	sqlDatabasesClient               sql.DatabasesClient
```
Given we're currently using the Riviera SDK for this functionality, this client isn't used. As such - is it possible to remove it for the moment?

tombuildsstuff commented on builtin/providers/azurerm/config.go:
```diff
@@ -99,6 +100,11 @@ type ArmClient struct {
+
+	sqlDatabasesClient               sql.DatabasesClient
+	sqlElasticPoolsClient            sql.ElasticPoolsClient
+	sqlRecommendedElasticPoolsClient sql.RecommendedElasticPoolsClient
```
Given we're currently using the Riviera SDK for this functionality, this client isn't used. As such - is it possible to remove it for the moment?

tombuildsstuff commented on builtin/providers/azurerm/config.go:
```diff
@@ -99,6 +100,11 @@ type ArmClient struct {
+	sqlDatabasesClient               sql.DatabasesClient
+	sqlElasticPoolsClient            sql.ElasticPoolsClient
+	sqlRecommendedElasticPoolsClient sql.RecommendedElasticPoolsClient
+	sqlServersClient                 sql.ServersClient
```
Given we're currently using the Riviera SDK for this functionality, this client isn't used. As such - is it possible to remove it for the moment?

dominik-lekse replied:
I have removed the unused clients except for `sqlElasticPoolsClient` which is used by the resource. I assume, Terraform has deprecated the Riviera SDK and will prefer the Azure RM Go SDK, right?

tombuildsstuff commented on builtin/providers/azurerm/resource_arm_sql_elasticpool.go:
```diff
@@ -0,0 +1,217 @@
+	d.Set("server_name", serverName)
+	d.Set("edition", string(elasticPool.Edition))
+	d.Set("dtu", int(*elasticPool.Dtu))
+	d.Set("db_dtu_min", int(*elasticPool.DatabaseDtuMin))
```
is it worth checking that `resp.ElasticPoolProperties` isn't `nil` before assigning these, to prevent a possible crash?

dominik-lekse replied:
Safe guard is now included.

dominik-lekse commented:
Thanks for the feedback @tombuildsstuff. With the recent commits, I addressed your comments. In addition, I have included the changes proposed in #12678 in this pull request to allow managing databases in elastic pools. Also, the test `TestAccAzureRMSqlDatabase_elasticPool` is now completed, i.e. it verifies creating a database in a new elastic pool using the new resource.

```
=== RUN   TestAccAzureRMSqlDatabase_elasticPool
--- PASS: TestAccAzureRMSqlDatabase_elasticPool (186.53s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/azurerm	186.553s
```

tombuildsstuff commented:
Tests pass:
<img width="411" alt="screen shot 2017-05-03 at 05 57 50" src="https://cloud.githubusercontent.com/assets/666005/25649003/9fadc448-2fc9-11e7-82a8-4ded62803b8c.png">
